core/src/main/java/com/linecorp/armeria/common/logging/ContentPreviewer.java

```diff
@@ -0,0 +1,110 @@
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.internal.ArmeriaHttpUtil;
+
+import io.netty.buffer.ByteBuf;
+
+/**
+ * Produces the preview of {@link RequestLog}.
+ */
+public interface ContentPreviewer {
+
+    /**
+     * Creates a new instance of {@link ContentPreviewer} which produces the preview through {@code reproducer}
+     * when the contents has been aggregated more than {@code length} bytes.
+     */
+    static ContentPreviewer ofBinary(int length, Function<? super ByteBuf, String> reproducer) {
+        return ofBinary(length, (headers, buffer) -> reproducer.apply(buffer));
+    }
+
+    /**
+     * Creates a new instance of {@link ContentPreviewer} which produces the preview through {@code reproducer}
+     * when the contents has been aggregated more than {@code length} bytes.
+     */
+    static ContentPreviewer ofBinary(int length,
+                                     BiFunction<? super HttpHeaders, ? super ByteBuf, String> reproducer) {
+        return ByteBufAggregatingPreviewer.create(length, reproducer);
+    }
+
+    /**
+     * Creates a new instance of {@link ContentPreviewer} which produces the text
+     * with the maximum {@code length} limit.
+     * @param length the maximum length of the preview.
+     * @param defaultCharset the default charset for a request/response with unspecified charset in
+     *                       {@code "Content-Type"} header.
+     */
+    static ContentPreviewer ofText(int length, Charset defaultCharset) {
+        checkArgument(length >= 0, "length : %d (expected: >= 0)", length);
+        if (length == 0) {
+            return disabled();
+        }
+        return new StringAggregatingPreviewer(length, defaultCharset);
+    }
+
+    /**
+     * Creates a new instance of {@link ContentPreviewer} which produces the text
+     * with the maximum {@code length} limit.
+     * @param length the maximum length of the preview.
+     */
+    static ContentPreviewer ofText(int length) {
+        return ofText(length, ArmeriaHttpUtil.HTTP_DEFAULT_CONTENT_CHARSET);
+    }
+
+    /**
+     * A dummy {@link ContentPreviewer} which discards everything it collected and produces {@code null}.
+     */
+    static ContentPreviewer disabled() {
+        return NoopContentPreviewer.INSTANCE;
+    }
+
+    /**
+     * Invoked after request/response headers is received.
+     */
+    void onHeaders(HttpHeaders headers);
+
+    /**
+     * Invoked after request/response data is received.
+     * Note that it is not invoked when the request/response is completed or {@link #isDone()} returns
+     * {@code true} even if a new content is received.
+     */
+    void onData(HttpData data);
+
+    /**
+     * Produces the preview of {@link RequestLog}.
+     * Note that it is invoked when the request or response is ended
+     * or the preview has been ready to be produced.
```

Comment:
Could you add `@return` which describes what it returns, e.g. `@return the preview, or {@code null} if ...`